utils.js

```diff
@@ -350,8 +359,91 @@ function _formatAttachment(attachment1, attachment2) {
         attachment1: attachment1,
         attachment2: attachment2
       };
+    case "MessageImage":
+      return {
+        type: "photo",
+        ID: blob.legacy_attachment_id,
+        filename: blob.filename,
+        thumbnailUrl: blob.thumbnail.uri,
+
+        previewUrl: blob.preview.uri,
+        previewWidth: blob.preview.width,
+        previewHeight: blob.preview.height,
+
+        largePreviewUrl: blob.large_preview.uri,
+        largePreviewWidth: blob.large_preview.width,
+        largePreviewHeight: blob.large_preview.height,
+
+        url: blob.large_preview.uri,
+        width: blob.original_dimensions.x,
+        height: blob.original_dimensions.y,
+      }
+    case "MessageAnimatedImage":
+      return {
+        type: "animated_image",
+        name: blob.filename,
+        facebookUrl: blob.animated_image.uri,
+        previewUrl: blob.preview_image.uri,
+        previewWidth: blob.preview_image.width,
+        previewHeight: blob.preview_image.height,
+        thumbnailUrl: blob.preview_image.uri,
+        ID: blob.legacy_attachment_id,
+        filename: blob.filename,
+        width: blob.animated_image.width,
+        height: blob.animated_image.height,
+        url: blob.animated_image.uri,
+        rawGifImage: blob.animated_image.uri,
+        animatedGifUrl: blob.animated_image.uri,
+        animatedGifPreviewUrl: blob.preview_image.uri,
+        animatedWebpUrl: blob.animated_image.uri,
+        animatedWebpPreviewUrl: blob.preview_image.uri,
+      };
+    case "StickerAttachment":
+        return {
+          type: "sticker",
+          url: blob.url,
+          stickerID: blob.id,
+          packID: blob.pack.id,
+          frameCount: blob.frame_count,
+          frameRate: blob.frame_rate,
+          framesPerRow: blob.frames_per_row,
+          framesPerCol: blob.frames_per_column,
+          spriteURI: blob.sprite_image,
+          spriteURI2x: blob.sprite_image_2x,
+          height: blob.height,
+          width: blob.width,
+          caption: blob.label,
+          description: blob.label
+        };
+    case "ExtensibleAttachment":
+      return {
+        type: "share",
+        description: blob.story_attachment.description.text,
+        ID: blob.legacy_attachment_id,
+        subattachments: blob.story_attachment.subattachments,
+        width: blob.story_attachment.media.image.width,
+        height: blob.story_attachment.media.image.height,
+        image: blob.story_attachment.media.image.uri,
+        playable: blob.story_attachment.media.is_playable,
+        duration: blob.story_attachment.media.playable_duration_in_ms,
+        source: blob.story_attachment.source.text,
+        title: blob.story_attachment.title_with_entities.text,
+        facebookUrl: blob.story_attachment.url,
+        target: blob.story_attachment.target.__typename,
```

Comment:
Are you sure this should be `__typename` here?

Reply:
Well generate some data and see if you think something else.  I had limited content sharing options to try for but it seemed the field that matched it best.   Either add additional debug lines or rename the case to something else and it will auto-dump the entire data stream when you share something.  If you can provide the json that also works:)

Reply:
Ok I just checked. When you share any link you'll get
```
"target": {
  "__typename": "ExternalUrl"
}
```

This indicates that there is such a thing as an "internal" URL, given that it's FB I tried sharing a URL to an FB event:
```
    "target": {
      "__typename": "Event",
      "eventID": "116707585671845",
      "guestStatus": null,
      "canViewerJoin": true,
      "timeString": "Friday, November 10 at 5:00pm",
      "event_place": {
        "name": "Dogpatch Boulders"
      },
      "social_context": {
        "text": "Fitness · 666 people"
      },
      "watchStatus": "UNWATCHED",
      "connection_style": "INTERESTED",
      "canViewerWatch": true
    },
```

I couldn't find other objects that had different fields, so let's do something like:
```
target: blob.story_attachment.target == null ? undefined : {
  type: blob.story_attachment.target.__typename,
  eventID: blob.story_attachment.target.eventID,
  guestStatus: blob.story_attachment.target.guestStatus,
  canViewerJoin: blob.story_attachment.target.canViewerJoin,
  timeString: blob.story_attachment.target.timeString,
  eventPlaceName: blob.story_attachment.target.event_place == null ? null : blob.story_attachment.target.event_place.name,
  socialContext: blob.story_attachment.target.social_context == null ? null : blob.story_attachment.target.social_context.text,
  watchStatus: blob.story_attachment.target.watchStatus,
  connectionStyle: blob.story_attachment.target.connection_style,
  canViewerWatch: blob.story_attachment.target.canViewerWatch
}
```

Reply:
Target before was not a complex structure (I don't believe) as such I would not suggest making it a complex structure.  I agree on providing more details however, so maybe add a new field with target_details with that structure?  What do you think.  Mainly I want to try and break as few as applications as possible that used the data before when updating the lib?

Reply:
As far as I remember, the target before was simply an unprocessed object from facebook and the closest imitation would probably be to simply use `blob.story_attachment.target`. Afaict there's no evidence that `target` was ever a string.

However, I think what @bsansouci is getting at is that it would be nice to start processing this object so that it can be more consistent. I'm not sure that this necessarily needs to be part of this fix, but you can definitely feel free to.
If we do this, I think we would need a switch on `blob.story_attachment.target.__typename` and do something like what @bsansouci is suggesting in `case "Event":`, with some exception or warning logging in the case where we don't recognize the `__typename`.
If we do this, then a section would probably need to be added under the *Attachments* section [here](https://github.com/Schmavery/facebook-chat-api/blob/master/DOCS.md#apilistencallback) for types of share. Technically the fields of the attachments should be documented a little more in the future but I won't make you do that 😛

Reply:
Ah, well then I have no problem making the object his object if people want that in the PR.

It would be good if someone setup a FB account with a bunch of test posts (or links to posts?) part of what I ran into an issue with testing the interface was finding all the content types to generate the structures.  As I mention in the PR, I didn't even see how to send video through the web interface to update that structure.

Reply:
In order to avoid too much back and forth on this PR (which can already help people in its current form 🎉 ), I'd love it if we could just change target to be `blob.story_attachment.target` and then merge this, followed up with a second PR for any additional fixes, if you have time/willingness :)